brushfire-core/src/main/scala/com/stripe/brushfire/Samplers.scala

```diff
@@ -60,40 +60,44 @@ case class RFSampler(numTrees: Int, featureRate: Double, samplingRate: Double =
   }
 }
 
-case class TimeGroupedSampler[K](base: Sampler[K], period: Long, groups: Int) extends Sampler[K] {
+case class TimeGroupedSampler[-M, -K](timestamp: M => Long, base: Sampler[M, K], period: Long, groups: Int) extends Sampler[M, K] {
   def numTrees = base.numTrees * groups
 
-  def timesInTrainingSet(id: String, timestamp: Long, treeIndex: Int) = {
-    val timeGroup = (timestamp / period) % groups
+  def timesInTrainingSet(metadata: M, treeIndex: Int) = {
+    val timeGroup = (timestamp(metadata) / period) % groups
     val treeGroup = treeIndex % groups
     val indexInGroup = treeIndex / groups
 
     if (timeGroup == treeGroup)
-      base.timesInTrainingSet(id, timestamp, indexInGroup)
+      base.timesInTrainingSet(metadata, indexInGroup)
     else
       0
   }
 
-  def includeInValidationSet(id: String, timestamp: Long, treeIndex: Int) = {
-    timesInTrainingSet(id, timestamp, treeIndex) == 0
+  def includeInValidationSet(metadata: M, treeIndex: Int) = {
+    timesInTrainingSet(metadata, treeIndex) == 0
   }
 
-  def includeFeature(name: K, treeIndex: Int, leafIndex: Int) = base.includeFeature(name, treeIndex, leafIndex)
+  def includeFeature(metadata: M, name: K, treeIndex: Int, leafIndex: Int) = {
+    base.includeFeature(metadata, name, treeIndex, leafIndex)
+  }
 }
 
-case class OutOfTimeSampler[K](base: Sampler[K], threshold: Long) extends Sampler[K] {
+case class OutOfTimeSampler[-M, -K](timestamp: M => Long, base: Sampler[M, K], threshold: Long) extends Sampler[M, K] {
```

Comment:
Or this could change to being a `MetadataFilterSampler` or something that takes a `M=>Boolean` rather than hardcoding the idea of a `Long` threshold.